File: tools/analyze_templates.py
import os
import zipfile
import re
import glob

TEMPLATE_DIR = r"C:\xampp\htdocs\clearpass\Certificate Templates"

def analyze_docx(file_path):
    try:
        with zipfile.ZipFile(file_path) as zf:
            xml_content = zf.read('word/document.xml').decode('utf-8')
            # Regex to find placeholders like {resident_name}
            # Handles simple cases. Complex cases might have XML tags inside the braces
            # e.g. {<w:t>resident</w:t>_<w:t>name</w:t>} - this is common in Word
            # So we first strip xml tags? Or use a more robust regex.
            
            # Simple approach: remove all XML tags then find {...}
            text_only = re.sub(r'<[^>]+>', '', xml_content)
            variables = re.findall(r'\{[a-zA-Z0-9_]+\}', text_only)
            return list(set(variables))
    except Exception as e:
        return [f"Error: {str(e)}"]

def main():
    print(f"Scanning directory: {TEMPLATE_DIR}")
    
    if not os.path.exists(TEMPLATE_DIR):
        print("Directory not found!")
        return

    all_files = os.listdir(TEMPLATE_DIR)
    docx_files = [f for f in all_files if f.lower().endswith('.docx')]
    doc_files = [f for f in all_files if f.lower().endswith('.doc')]
    
    print("\n--- File Summary ---")
    print(f"Total files: {len(all_files)}")
    print(f".docx files (Compatible): {len(docx_files)}")
    print(f".doc files (Incompatible - Need Conversion): {len(doc_files)}")
    
    print("\n--- Incompatible Files (.doc) ---")
    for f in doc_files:
        print(f" - {f}")

    print("\n--- Variable Analysis (.docx) ---")
    for f in docx_files:
        path = os.path.join(TEMPLATE_DIR, f)
        vars_found = analyze_docx(path)
        print(f"\nFile: {f}")
        if vars_found:
            print(f"  Variables: {', '.join(vars_found)}")
        else:
            print("  No variables found (or parsing failed)")

if __name__ == "__main__":
    main()

File: analyze_templates.js
import fs from 'fs';
import path from 'path';
import AdmZip from 'adm-zip';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const TEMPLATE_DIR = path.join(__dirname, 'Certificate Templates');

function extractPlaceholders(filePath) {
    try {
        const zip = new AdmZip(filePath);
        const contentXml = zip.readAsText("word/document.xml");
        
        // Remove XML tags to get plain text
        const plainText = contentXml.replace(/<[^>]+>/g, '');
        
        // Find {variable} or {{variable}}
        // The previous regex was a bit loose. Let's try to capture anything inside braces.
        const matches = plainText.match(/\{+[^{}]+\}+/g) || [];
        
        return [...new Set(matches)]; // Unique
    } catch (e) {
        console.error(`Error reading ${filePath}:`, e.message);
        return [];
    }
}

function analyzeTemplates() {
    if (!fs.existsSync(TEMPLATE_DIR)) {
        console.error("Template directory not found:", TEMPLATE_DIR);
        return;
    }

    const files = fs.readdirSync(TEMPLATE_DIR).filter(f => f.endsWith('.docx'));
    const analysis = {};
    const allPlaceholders = new Set();

    console.log("Analyzing templates...\n");

    files.forEach(file => {
        const filePath = path.join(TEMPLATE_DIR, file);
        const placeholders = extractPlaceholders(filePath);
        analysis[file] = placeholders;
        placeholders.forEach(p => allPlaceholders.add(p));
        
        console.log(`[${file}]`);
        if (placeholders.length > 0) {
            console.log(`  Found: ${placeholders.join(', ')}`);
        } else {
            console.log(`  No placeholders found (or parsing failed)`);
        }
    });

    console.log("\n--- SUMMARY OF ALL PLACEHOLDERS ---");
    console.log([...allPlaceholders].sort().join('\n'));
}

analyzeTemplates();
